java-frontend/src/main/java/org/sonar/java/ast/visitors/CognitiveComplexityVisitor.java

```diff
@@ -0,0 +1,282 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.ast.visitors;
+
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.java.resolve.SemanticModel;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.BreakStatementTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.CompilationUnitTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.ContinueStatementTree;
+import org.sonar.plugins.java.api.tree.DoWhileStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForEachStatement;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.SwitchStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.sonar.plugins.java.api.tree.Tree.Kind.CONDITIONAL_AND;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.CONDITIONAL_OR;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.IF_STATEMENT;
+
+public class CognitiveComplexityVisitor extends BaseTreeVisitor {
+  private final List<JavaFileScannerContext.Location> locations;
+  private final Set<Tree> ignored;
+  private int complexity;
+  private int nesting;
+  private boolean ignoreNesting;
+
+  private CognitiveComplexityVisitor() {
+    complexity = 0;
+    nesting = 1;
+    ignoreNesting = false;
+    locations = new ArrayList<>();
+    ignored = new HashSet<>();
+  }
+
+  public static JavaFileScannerContext.CognitiveComplexity methodComplexity(@Nullable SemanticModel semanticModel, MethodTree methodTree) {
+    if (shouldAnalyzeMethod(semanticModel, methodTree)) {
+      CognitiveComplexityVisitor visitor = new CognitiveComplexityVisitor();
+      methodTree.accept(visitor);
+      return new JavaFileScannerContext.CognitiveComplexity(visitor.complexity, visitor.locations);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this instantiation of "CognitiveComplexity". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2440)